src/transport/TcpTransport.cpp

```diff
@@ -277,6 +281,10 @@ void TcpTransport::readNextMessageIntCallback(BufferEvent* event, TcpTransport*
 
       transport->messageReceived(msg, event->getPeerAddrPort());
     }
+    /**
+     * reset read water mark to 4
+     */
+    event->setWatermark(EV_READ, 4, 0);
```

Comment:
Move to line 259.

Reply:
at line 259, the water matk is already 4
so I think need reset read water mark to 4 after received whole data and user's callback

Reply:
If `needed` is greater than 0, `hdr` is incomplete, return and wait for more data. The function does not return here.